File: Queue/Queue_singlyLL.c
/*
** Author: Bryce Calhoun
** Description: Queue ADT implemented with a singly liked list.
*/

#include "singlyLL.c"
#include <stdio.h>
#include <stdlib.h>

struct Queue{
    /*
    Description: Queue structure defined with an integer variable to track the size
                and a singlyLL object as the underlying data structure
    */
    int size;
    struct singlyLL* queue;
};


struct Queue* create_queue(){
    /*
    Description: Function to allocate memory for a queue object, initialize its attributes, and 
                return a pointer to it
    */
    struct Queue* q = malloc(sizeof(struct Queue));
    q->queue = create_list();
    q->size = 0;
};

int queue_is_empty(struct Queue* q){
    /*
    Description: Function to return 1(true) if the queue is empty and 0(false) otherwise
    */
    if(q->size == 0){
        return 1;
    }
    return 0;
};

void enqueue(struct Queue* q, void* value){
    /*
    Description: Function to take a void pointer and enqueue to the end of the queue
    */
    insert_back(q->queue, value);
    q->size++;
};

void* dequeue(struct Queue* q){
    /*
    Description: Function to remove a value from the front of the queue and return it
    */
    int empty = queue_is_empty(q);
    if(empty){
        printf("EMPTY QUEUE");
        return NULL;
    }
    void* value = remove_front(q->queue);
    q->size--;
    return value;
};

void* peek(struct Queue* q){
    /*
    Description: Function to read a value from the front of the queue without removing it
    */
    void* value = get_front(q->queue);
    return value;
};

File: Queue/Queue_singlyLL.py
"""
AUTHOR: Bryce Calhoun
DESCRIPTION: Queue ADT implemented with a singly linked list of node objects
"""

class Node:
    """
    DESCRIPTION: Object definition for a single node in the Linked List
    """
    def __init__(self, value, nextNode):
        self._value = value
        self._nextNode = nextNode

    def get_value(self):
        return self._value

    def get_next(self):
        return self._nextNode

    def set_next(self, node):
        self._nextNode = node
        return


class LinkedList:
    """
    DESCRIPTION: Object definition for the linked list holding the nodes
    """

    def __init__(self):
        self._front = None
        self._rear = None
        self._size = 0


    def get_size(self):
        return self._size

    def get_front(self):
        return self._front

    def set_front(self, node):
        self._front = node
        return

    def set_rear(self, node):
        self._rear = node
        return

    def get_rear(self):
        return self._rear

    def is_empty(self):
        if self.get_size() == 0:
            return True
        return False

    def insert_rear(self, value):

        if self.is_empty():
            newNode = Node(value, None)
            self.set_front(newNode)
        else:
            newNode = Node(value, None)
            self.get_rear().set_next(newNode)

        self.set_rear(newNode)
        self._size += 1
        return

    def remove_front(self):
        if self.is_empty():
            return

        returnNode = self.get_front()

        self.set_front(returnNode.get_next())

        self._size -= 1

        if self.get_size() == 0:
            self.set_front(None)
            self.set_rear(None)

        return returnNode


class Queue_singlyLL:
    """
    DESCRIPTION: Object definition for the queue ADT, implemented with a singly linked list
    """

    def __init__(self):
        self._size = 0
        self._list = LinkedList()

    def get_size(self):
        return self._size

    def is_empty(self):
        if self.get_size() == 0:
            return True
        return False


    def enqueue(self, value):

        self._list.insert_rear(value)
        self._size += 1
        return

    def dequeue(self):

        if self.is_empty():
            return

        returnNode = self._list.remove_front()
        self._size -= 1

        return returnNode

    def peek(self):
        return self._list.get_front()


